Add docs for Tail-based sampling with the EDOT collector (#10022)
elastic/elastic-agent: docs/reference/edot-collector/config/tail-based-sampling.md

To enforce a specific order of calculations and sampling decisions in the EDOT Collector, you can use the [Forward connector](https://github.com/open-telemetry/opentelemetry-collector/tree/main/connector/forwardconnector). Split the traces pipeline in two steps using the connector, with the first part applying calculations and the second part applying the tail-based sampling decision.

## Create a two-step trace pipeline

Configure a two-step trace pipeline, ensuring that the first step includes the `elasticapm` connector and `forward` connector under `exporters` section, and that the second step includes the `tail_sampling` processor.

```yaml
connectors:
  elasticapm: {} # Elastic APM Connector
  forward: # Forward connector necessary for tail-based sampling

processors:
  tail_sampling:
    decision_wait: 5s
    num_traces: 50000
    expected_new_traces_per_sec: 10
    policies:
      - name: sample_10_percent
        type: probabilistic
        probabilistic:

>           sampling_percentage: 10

Reviewer: Should we include the `probabilistic` sampler as a default config? Ideally, TBS is expensive and the users should use HBS if they only want to use probailisitic sampler. Personally, I think we should not have an example to encourage this.

Reviewer: This is also captured in upstream README.

service:
  pipelines:
    traces/1-process-elastic:
      receivers: [ otlp ]
      processors: [ elastictrace ]
      exporters: [ elasticapm, forward ]
    traces/2-process-tbs:
      receivers: [ forward ]
      processors: [ tail_sampling ]
      exporters: [ elasticsearch/otel ]
```

## Configuration for a load-balancing Collector

> When using a load-balancing Collector with downstream collectors, all traces must go to the load-balancing Collector for the sampling decision to happen. Set the `traceID` as the routing key in the load-balancing Collector so that tail-sampling is applied to the downstream collectors.

Reviewer 2: ```suggestion
To horizontally scale collectors with tail sampling enabled, all traces should go through a load-balancing collector in front of the downstream tail-sampling collectors. Set the `traceID` as the routing key in the load-balancing Collector so that all traces belonging to the same trace ID go to the same downstream tail-sampling collector.
```

> To horizontally scale collectors with tail sampling enabled, all traces should go through a load-balancing collector in front of the downstream tail-sampling collectors. Set the `traceID` as the routing key in the load-balancing Collector so that all traces belonging to the same trace ID go to the same downstream tail-sampling collector.

Author: @theletterf maybe you want to review this again after @carsonip's suggested change (that I committed).

Reviewer 3: Edited!

---
navigation_title: Tail-based Sampling
description: Configuration of the EDOT Collector for Tail-based sampling.
applies_to:
stack:
serverless:
observability:
product:
edot_collector: ga
products:
- id: cloud-serverless
- id: observability
- id: edot-collector
---

# Configure tail-based sampling

> Tail-based sampling is where the decision to sample a trace takes place by considering all or most of the spans within the trace. This is in contrast to Head-based sampling, which is on the client-side where traces are created.
> 
> Because metrics are calculated and aggregations are performed on trace data in the EDOT collector, some specific configuration is required to ensure that the tail sampling decisions are made after these calculations.
> Otherwise, the metrics and aggregations would be incorrect due to the partial representation of traces.
> 

Reviewer 3: ```suggestion
Tail-based sampling analyzes a complete trace before deciding whether to keep it, enabling intelligent decisions based on factors like errors or high latency. This is superior to head-based sampling, which makes a premature decision at the start of a trace.

Within the OpenTelemetry Collector, any processor that generates metrics from traces must run before the tail-sampling processor. If sampling happens first, metrics will be calculated on an incomplete data set, leading to inaccurate and misleading reporting.
```

Author: I wouldn't use the word "superior" because it really depends on the use case. In some cases, probabilistic sampling on the client side is the right choice for the scenario.

Reviewer 3: Agreed, edited to make it more neutral.

> Enforcing a specific order of calculations and sampling decisions can be accomplished in the EDOT Collector using the [forward connector](https://github.com/open-telemetry/opentelemetry-collector/tree/main/connector/forwardconnector). The traces pipeline should be split into two parts using the forward connector, with the first part applying calculations and the second part applying the tail-based sampling decision.

Reviewer 3: ```suggestion
To enforce a specific order of calculations and sampling decisions in the EDOT Collector, you can use the [Forward connector](https://github.com/open-telemetry/opentelemetry-collector/tree/main/connector/forwardconnector). Split the traces pipeline in two steps using the connector, with the first part applying calculations and the second part applying the tail-based sampling decision.
```

> ## Configure the EDOT Collector to have a two-part trace pipeline

Reviewer 3: ```suggestion
## Create a two-step trace pipeline
```

Configure a two-part trace pipeline, ensuring that the first one includes the `elasticapm` exporter and `forward` exporter and the second one includes the `tail_sampling` processor.

:::{tab-item} EDOT Collector config
```yaml
processors:
  tail_sampling:
    decision_wait: 5s
    num_traces: 50000
    expected_new_traces_per_sec: 10
    policies:
      - name: sample_10_percent
        type: probabilistic
        probabilistic:
          sampling_percentage: 10
service:
  pipelines:
    traces/1:
      receivers: [ otlp ]
      processors: [ elastictrace ]
      exporters: [ elasticapm, forward ]
    traces/2:
      receivers: [ forward ]
      processors: [ tail_sampling ]
      exporters: [ elasticsearch/otel ]
```

> :::

Author: Sure, what do you think about `traces/process-elastic`, `traces/process-tbs`?

## Configure a load-balancing EDOT Collector with downstream collectors

> All traces must go to the same collector in order for the sampling decision to be made. Therefore, when using a load-balancing collector with downstream collectors, the `traceID` should be set as the routing key in the load-balancing collector and tail-sampling should be applied in the downstream collectors.

Reviewer 3: Correct!
